src/table/iterator.rs

```diff
@@ -189,6 +202,267 @@ impl BlockIterator {
+        match self.table.as_ref().block(self.bpos, self.use_cache()) {
+            Ok(block) => {
+                let block_iterator = self.get_block_iterator(block);
+                block_iterator.seek_to_first();
+                self.err = block_iterator.err.clone();
+            }
+            Err(err) => self.err = Some(err.into()),
+        }
+    }
+
+    pub fn seek_to_last(&mut self) {
+        let num_blocks = self.table.as_ref().offsets_length();
+        if num_blocks == 0 {
+            self.err = Some(IteratorError::EOF);
+            return;
+        }
+        self.bpos = num_blocks - 1;
+        match self.table.as_ref().block(self.bpos, self.use_cache()) {
+            Ok(block) => {
+                let block_iterator = self.get_block_iterator(block);
+                block_iterator.seek_to_last();
+                self.err = block_iterator.err.clone();
+            }
+            Err(err) => self.err = Some(err.into()),
+        }
+    }
+
+    fn seek_helper(&mut self, block_idx: usize, key: &Bytes) {
+        self.bpos = block_idx;
+        match self.table.as_ref().block(self.bpos, self.use_cache()) {
+            Ok(block) => {
+                let block_iterator = self.get_block_iterator(block);
+                block_iterator.seek(key, SeekPos::Origin);
+                self.err = block_iterator.err.clone();
+            }
+            Err(err) => self.err = Some(err.into()),
+        }
+    }
+
+    fn seek_from(&mut self, key: &Bytes, whence: SeekPos) {
+        self.err = None;
+        match whence {
+            SeekPos::Origin => self.reset(),
+            _ => {}
+        }
+
+        let idx = util::search(self.table.as_ref().offsets_length(), |idx| {
+            use std::cmp::Ordering::*;
+            let block_offset = self.table.as_ref().offsets(idx).unwrap();
+            match COMPARATOR.compare_key(&block_offset.key, &key) {
+                Less => false,
+                _ => true,
+            }
+        });
+
+        if idx == 0 {
+            self.seek_helper(0, key);
+            return;
+        }
+
+        self.seek_helper(idx - 1, key);
+        if IteratorError::check_eof(&self.err) {
+            if idx == self.table.as_ref().offsets_length() {
+                return;
+            }
+            self.seek_helper(idx, key);
+        }
+    }
+
+    // seek_inner will reset iterator and seek to >= key.
+    fn seek_inner(&mut self, key: &Bytes) {
+        self.seek_from(key, SeekPos::Origin);
+    }
+
+    // seek_for_prev will reset iterator and seek to <= key.
+    fn seek_for_prev(&mut self, key: &Bytes) {
+        self.seek_from(key, SeekPos::Origin);
+        if self.key() != key {
+            self.prev_inner();
+        }
+    }
+
+    fn next_inner(&mut self) {
+        self.err = None;
+
+        if self.bpos > self.table.as_ref().offsets_length() {
+            self.err = Some(IteratorError::EOF);
+            return;
+        }
+
+        if self.block_iterator.is_none() {
```

Comment:
Should also check if data is empty.

Reply:
In this implementation, `block_iterator` is always initialized with data. I guess it is not necessary.

Reply:
Then it should stack over flow when executing L373.

Reply:
Got it. I'll fix this soon.

Reply:
What about always giving the iterator an initial value? (e.g. points to the first block), then we won't need to deal with those Option<Iterator>s

Reply:
I think there always be an `Option`, the different is whether the `Option` is pushed down. Though I'm OK with your proposal.

Reply:
Now we will load the block either the iterator isn't initialized, or the data is empty.